File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    let nsl = getNsl(heights);
    let nsr = getNsr(heights);
    let finalArea = 0;
    for(let i = 0;i<heights.length;i++){
        let width = nsr[i]-nsl[i] -1;
        let newArea = heights[i]*width;
        finalArea = Math.max(finalArea,newArea);
    }
    return finalArea;
    
};

var getNsl = function(heights){
    let st = [];
    let res = [];
    for(let i =0;i<heights.length;i++){
        while(st.length!==0 && heights[st[st.length-1]]>= heights[i]){
            st.pop();
        }
        res[i] = st.length === 0 ? -1:st[st.length - 1];
        st.push(i);
    } 
    return res;
}

var getNsr = function(heights){
    let st = [];
    let res = [];
    for(let i = heights.length;i>=0;i--){
        while(st.length!==0 && heights[st[st.length-1]]> heights[i]){
            st.pop();
        }
       res[i] = st.length === 0 ? heights.length:st[st.length - 1];
        st.push(i);
    } 
    return res;
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:

    def getNsl(self,heights):
        n = len(heights)
        stack = []
        res = [0]*n
        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            res[i] = -1 if not stack else stack[-1]
            stack.append(i)
        return res

    def getNsr(self,heights):
        n = len(heights)
        stack = []
        res = [0]*n
        for i in range(n-1,-1,-1):
            while stack and heights[stack[-1]] > heights[i]:
                stack.pop()
            res[i] = n if not stack else stack[-1]
            stack.append(i)
        return res

    def largestRectangleArea(self, heights: List[int]) -> int:
        nsl = self.getNsl(heights)
        nsr = self.getNsr(heights)
        print("nsl",nsl)
        print("nsr",nsr)
        n = len(heights)
        finalArea = 0
        for i in range(n):
            width = nsr[i] - nsl[i] - 1
            area = heights[i]*width
            finalArea = max(area,finalArea)
        return finalArea


        